src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,421 @@
+void AccountValidator::checkNextHost()
+{
+    if (mCurrentServiceTested == SERVICE_IMAP) {
+        if (mCurrentServiceIndex < mImapServices->count()) {
+            
+            if (mImapSession == NULL) {
+                mImapSession = new IMAPAsyncSession();
+                mImapSession->setUsername(mUsername);
+                mImapSession->setPassword(mPassword);
+            }
+            
+            mImapServer = (NetService *) mImapServices->objectAtIndex(mCurrentServiceIndex);
+            mImapSession->setHostname(mImapServer->hostname());
+            mImapSession->setPort(mImapServer->port());
+            mImapSession->setConnectionType(mImapServer->connectionType());
+            
+            mOperation = (IMAPOperation *)mImapSession->checkAccountOperation();
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+        
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else if (mCurrentServiceTested == SERVICE_POP){
+        if (mCurrentServiceIndex < mPopServices->count()) {
+            POPAsyncSession *popSession = new POPAsyncSession();
+            popSession->setUsername(mUsername);
+            popSession->setPassword(mPassword);
+            
+            mPopServer = (NetService *) mPopServices->objectAtIndex(mCurrentServiceIndex);
+            popSession->setHostname(mPopServer->hostname());
+            popSession->setPort(mPopServer->port());
+            popSession->setConnectionType(mPopServer->connectionType());
+            
+            mOperation = (POPOperation *)popSession->checkAccountOperation();
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else if (mCurrentServiceTested == SERVICE_SMTP){
+        if (mCurrentServiceIndex < mSmtpServices->count()) {
+            SMTPAsyncSession *smtpSession = new SMTPAsyncSession();
+            smtpSession->setUsername(mUsername);
+            smtpSession->setPassword(mPassword);
+            
+            mSmtpServer = (NetService *) mSmtpServices->objectAtIndex(mCurrentServiceIndex);
+            smtpSession->setHostname(mSmtpServer->hostname());
+            smtpSession->setPort(mSmtpServer->port());
+            smtpSession->setConnectionType(mSmtpServer->connectionType());
+            
+            mOperation =  (SMTPOperation *)smtpSession->checkAccountOperation(Address::addressWithMailbox(mEmail));
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else {
+        MC_SAFE_RELEASE(mPopSession);
```

Comment:
I had purposely left it out because you mentioned earlier about keeping the IMAPSession because it might be useful. Should I just not think about that for now?